### ISSUE TITLE
Support threads display

### ISSUE THREAD

**Issue by toby63** [Nheko-Reborn/nheko#1273]
### The Problem

Matrix introduced threads inside rooms a while ago and while Nheko displays the content, there are no real threads.

I have seen different approaches so far:
- Threads as a second or third seperate section (seen as an option in Element)
- Threads as a seperate "room", where you click on a thread's "title" (first message) and then a new "room" opens (seen in Element for Android)

### The Solution

I like two ideas:
- threads as a second section
- threads as minimized message: when you click on them, they open fully and indented and you can write in them (maybe with a seperate editor inside)

### Alternatives

There are probably more design ideas.

### Happens in the latest version

- [X] Yes, this feature is still missing.

**Comment by deepbluev7** [comment]
master branch already supports threads.

**Comment by toby63** [comment]
> master branch already supports threads.

I use a very recent master version (from yesterday or so).
Now that you said it, I noticed an option in the right-click menu that opens threads (as a seperate section), but this is only half the implementation IMHO.
Because in the regular view, you can't see that something is a thread, but I assume that that was exactly the idea upstream had intended, that you could have some kind of seperation and directly see different "topics"/threads and keep content seperated, instead of having the regular stream of content.

**Comment by deepbluev7** [comment]
No, you can see, that something is in a thread. It has a colored piece of yarn next to it, if it is from a thread. This implementation is intentional, because I think it makes them more useful. It would take no effort to hide them, but it just would make it more annoying to use.

**Comment by toby63** [comment]
> No, you can see, that something is in a thread. It has a colored piece of yarn next to it, if it is from a thread. This implementation is intentional, because I think it makes them more useful. It would take no effort to hide them, but it just would make it more annoying to use.

Is there a chance you would add a seperate option?

My idea for this option is:
- only show the first message of the thread
- mark this message with a more visible icon (the icon now is barely visible IMHO)
    the icon could be directly next to the text
- show an icon with a number, to indicate the number of messages included in it
   - maybe also indicating the number of new messages
- the thread is then opened when you click on this message

**Comment by deepbluev7** [comment]
Probably not. I have no interest in using such an option, so it is unlikely I would add it.

**Comment by toby63** [comment]
> Probably not. I have no interest in using such an option, so it is unlikely I would add it.

To clarify, I of course accept it that you don't want to implement something, because this is FOSS and you can do what you want etc.,
...even though in this case I would argue that my described way is what upstream intended and your way is basically avoiding this feature because you don't like it...

Anyway one thing to think about is the second part of your answer:
The basic reason for having options in software is to have... well... other options than the ones defined by default (probably aka "the options you like").
So you basically argue that you don't want to provide any options, because you only use the default way and don't want anyone to be able to use it in a different way. That is an "interesting" way of seeing things.

**Comment by deepbluev7** [comment]
No, I never said someone else can't open a PR with such a feature. You asked "Is there a chance you would add a separate option?", and to that the answer is no, I won't add it and I can't imagine any of the core contributors would. If someone wants to spend their time to build something they want to have, I would encourage that and even merge it, if it doesn't conflict with other stuff.

Also, there is no upstream. Matrix is a specification and it does not specify one specific way to use a feature. Several people complained about the implementation in Element and I built something different, that at least some of the complainers prefer. It is not avoiding it, but it is making the best out of its technical limitations until those are removed and then we can revisit it.

**Comment by toby63** [comment]
> No, I never said someone else can't open a PR with such a feature. You asked "Is there a chance you would add a separate option?", and to that the answer is no, I won't add it and I can't imagine any of the core contributors would. If someone wants to spend their time to build something they want to have, I would encourage that and even merge it, if it doesn't conflict with other stuff.

Ok, that sounds fair.
Could this be reopened then?
Or should I open a seperate issue?

> Also, there is no upstream. Matrix is a specification and it does not specify one specific way to use a feature. Several people complained about the implementation in Element and I built something different, that at least some of the complainers prefer. It is not avoiding it, but it is making the best out of its technical limitations until those are removed and then we can revisit it.

Well I don't know enough about the Matrix protocol, but IIRC the definition of a standard specification is having a detailed description of the technical and functional concept and even usecases for it. But if it is different here, I am no expert.
I only see what I see, which is, to use a different example:
_Imagine having a feature that makes text bolder; now your way of implementation is:
the text is not displayed bolder by default, it is displayed normal (as it was displayed before), but there is a little icon indicating that the text should be bold and when you click on that, this specific text is displayed bold._
This is not really a proper implementation of the feature to me, but this is of course just my opinion.
The "proper" way would be, to show the text bold by default and (if you want to) offer an option that disables bold text, if you don't like bold text.

But I want to say: I am not here to make you angry, but I am a bit confused, as I never had this situation of having a software that implements a specific protocol (which I called "upstream", which might be wrong) and the devs (you obviously) say, "we don't like this feature so we change it to our likings" and don't offer the "official" way.

> technical limitations

Could you elaborate that?
I am interested, what you would like to see.

**Comment by toby63** [comment]
For context: https://spec.matrix.org/v1.5/client-server-api/#threading

Quote:

> Threads allow users to visually branch their conversations in a room. Typically mostly used when a room is discussing multiple topics, threads provide more organisation of communication that traditional [rich replies](https://spec.matrix.org/v1.5/client-server-api/#rich-replies) can’t always offer.

> Clients SHOULD render threads differently to regular messages or replies in the timeline, such as by providing some context to what is going on in the thread but keeping the full conversation history behind a disclosure.

**Comment by toby63** [comment]
@deepbluev7 
It seems that your implementation is contrary to what the Specs say (Quotes from above):
1. >to visually branch their conversations in a room

   The conversation is not branched.
2. >Clients SHOULD render threads differently to regular messages or replies in the timeline

   The threads are not rendered differently, they only have a symbole that is very hard to even recognize.

2.a 
  >such as by providing some context to what is going on in the thread 

   There is no context or anything.

3. >but keeping the full conversation history behind a disclosure.

   The full conversation is not hidden or seperated.

**Comment by deepbluev7** [comment]
I don't see how what we do is different from the spec says. You need to click a button to see the whole thread, otherwise you just see pieces of it. Spec doesn't say anything about the main timeline not being allowed to show any thread messages and it is also a "SHOULD" in the spec, so clients can disregard it and implement it differently.

And in general, there is no reason to have different clients if they don't take different choices on how to implement certain features. If you are looking for a client with different choices, just pick a different client. If you prefer the way Element implements it, just use Element?

**Comment by toby63** [comment]
> I don't see how what we do is different from the spec says. You need to click a button to see the whole thread, otherwise you just see pieces of it. Spec doesn't say anything about the main timeline not being allowed to show any thread messages and it is also a "SHOULD" in the spec, so clients can disregard it and implement it differently.

You shouldn't be a lawyer otherwise you would be in huge trouble.
[I named all things already](https://github.com/Nheko-Reborn/nheko/issues/1273#issuecomment-1396082451), you are simply ignoring explicit formulations and obvious interpretations.
And I think you know that, but you ignore it because you don't like it.

If you would make a completely different software you could do everything you want, but if you follow a specific protocol you "should" (pun intended) implement this protocol.

"Should" does not mean you can ignore things.
Should means that you should achieve the stated goals, if possible.

> And in general, there is no reason to have different clients if they don't take different choices on how to implement certain features. If you are looking for a client with different choices, just pick a different client. If you prefer the way Element implements it, just use Element?

I guess you already know that the main reason people search for other clients than Element is not primarily the design or functionality (which is good if not very good), instead it is the performance and security concerns because of Electron and all that stuff.
Sadly most clients are "alpha" or "beta" status at best.

Also no it is not about having huge differences, clients which support specific Protocols or Platforms should at best support all features, they may offer additional options and features, but they should not scrap options and features.

**Comment by deepbluev7** [comment]
@redsky17 and I implement threads how we want it. We won't implement it differently, because we dislike how Element does it. So even if you find something in the spec, that says how it should look, it doesn't matter. We develop this client not to clone Element with better performance, but to make a client that fits our goals for a chat client. If you don't like that, there are ways to fix that, but we won't implement threads in a way we don't like.

**Comment by toby63** [comment]
> @redsky17 and I implement threads how we want it. We won't implement it differently, because we dislike how Element does it. So even if you find something in the spec, that says how it should look, it doesn't matter. We develop this client not to clone Element with better performance, but to make a client that fits our goals for a chat client. If you don't like that, there are ways to fix that, but we won't implement threads in a way we don't like.

This breaks the protocol.
You can keep your solution, but only as an additional option.
Not providing the main option is breaking the protocol.
I don't even understand what there is not to like about it, its simply essentially what threads are about.

**Comment by redsky17** [comment]
I think you are fundamentally misunderstanding what 'protocol' means with relation to matrix.  'breaking the protocol' would mean we support an entirely different client/server structure for threads that is incompatible with the matrix specification.  This is not the case with nheko, where we fully support the messages sent from synapse in the same structure as every other client that implements threads does.

Finally, regarding what the client / server api document actually states:
> Clients SHOULD render threads differently to regular messages or replies in the timeline
Nheko does this via showing thread replies with a marker/button denoting that they are part of a thread.

> keeping the full conversation history behind a disclosure.
Nheko does this via showing the entire context of the thread when you click on the thread marker/button on a message.

There are a some problems I've seen with the element-style and slack-style thread implementation that lead to usability issues that the nheko implementation does not have:

1. If you don't already have the thread view open somewhere (both element and slack accomplish this with a little sidebar view), there's no easy way to get back to the thread without searching through possibly days or weeks or however long of other messages to find it.
2. If you aren't directly mentioned in the thread or otherwise don't have notifications enabled for threads, you will be completely unaware of new messages being added to the thread

As an aside, 'should' in legal speak does not mean 'must.'  The words that mean 'must' are 'shall' and 'must.'

**Comment by toby63** [comment]
> I think you are fundamentally misunderstanding what 'protocol' means with relation to matrix. 'breaking the protocol' would mean we support an entirely different client/server structure for threads that is incompatible with the matrix specification. This is not the case with nheko, where we fully support the messages sent from synapse in the same structure as every other client that implements threads does.

I think you are nitpicking on definitions, while you actually would understand that by protocol I do not mean the technical implementation in the background, if you have read what I wrote.

> Finally, regarding what the client / server api document actually states:
> 
> > Clients SHOULD render threads differently to regular messages or replies in the timeline
> 
> Nheko does this via showing thread replies with a marker/button denoting that they are part of a thread.

As I said, barely visible. So it is not a good implementation.

> > keeping the full conversation history behind a disclosure.
> 
> Nheko does this via showing the entire context of the thread when you click on the thread marker/button on a message.

The conversation is not behind a disclosure, thats a fact.
 
> There are a some problems I've seen with the element-style and slack-style thread implementation that lead to usability issues that the nheko implementation does not have:
> 
>     1. If you don't already have the thread view open somewhere (both element and slack accomplish this with a little sidebar view), there's no easy way to get back to the thread without searching through possibly days or weeks or however long of other messages to find it.
> 
>     2. If you aren't directly mentioned in the thread or otherwise don't have notifications enabled for threads, you will be completely unaware of new messages being added to the thread

As I said, you can keep your way of implementation as an optional alternative, if these problems bother you.
Also you could report that upstream and ask for changes.

> As an aside, 'should' in legal speak does not mean 'must.' The words that mean 'must' are 'shall' and 'must.'

I never said that, but "should" does also not mean "optional" or "up to you" :wink:.

**Comment by deepbluev7** [comment]
> I never said that, but "should" does also not mean "optional" or "up to you".

Last time I asked the spec people, it did...

Anyway, in the end I don't care, I am happy with the current implementation and if it bothers you, then you can do something about it by either using a client with your preferred UX or patching your Nheko to do it like you want. :)

**Comment by toby63** [comment]
> > I never said that, but "should" does also not mean "optional" or "up to you".
> 
> Last time I asked the spec people, it did...
> 
> Anyway, in the end I don't care, I am happy with the current implementation and if it bothers you, then you can do something about it by either using a client with your preferred UX or patching your Nheko to do it like you want. :)

It's sad that the intended and formulated way in the specification is ignored by a project that by using the specification claims to follow it.

**Comment by turt2live** [comment]
👋 Spec Core Team (SCT) member for Matrix here - I hear this issue is questioning whether Nheko's approach to threads is legal under the spec.

Short version is **Nheko is completely legal in its approach for UI here**. 

Long version is the spec ([ref](https://spec.matrix.org/v1.5/client-server-api/#threading)), as already quoted, uses language that gives clients the ability to decide how exactly they want to represent threads. In fact, the spec *actively avoids* prescribing exact UI/UX behaviour for clients as that limits the usefulness of the protocol (also it's a protocol, not a UI library - it has no business dealing with UI). When the spec uses `SHOULD` (in either all caps or in lowercase), it means it's *suggesting* the client do something, but is *not* requiring the client do that.  Where a requirement is being communicated, it will be a `MUST`.

**Comment by turt2live** [comment]
For further clarity, since the sentence appears in the discussion a lot, "Threads allow users to visually branch their conversations in a room" is simply a feature description, not a specification. Rendering threads how Nheko does would also be counted as "branching", therefore even if the description was treated as specification, Nheko would still be perfectly in line with the spec.

**Comment by hiiamboris** [comment]
@deepbluev7

You've built the only usable desktop client so far, and one that respects OS theme (many thanks for that! saves me the trouble of implementing it myself :) But that also means you can't get away with "that's how I like it" answer, for now your userbase is big whether you like it or not ;) Plus even when you're not starting threads, you're bound to look at them when someone else starts.

I'm not going to promote the moronic implementation of Element (which is the same as it was in Gitter), where you have to sometimes scroll up 10+ screens to find where someone left an answer. Threads UI/UX is a historical PITA, and Nheko implementation is probably the best one so far. But I would like it to be better still. 

First, look at the UX shortcomings with me please:

1. I can't even tell the threads apart by their yarn color:

   ![](https://i.gyazo.com/45673dc41be3b4e73ab8652d40e5cffe.png)

   Last two messages belong to different threads but visually I don't see that, since the colors are too close.

2. It makes it easy to do the wrong thing: write the answer into the global conversation.

   Because the yarn icon is so tiny and is located at the far right side (not where eyes of the reader are likely to go), it looks like it belongs to the main conversation, and facilitates answers into it. But others will see this message as out of place and will be confused for sure. 

3. There's no connection between the thread topic and the answer, which complicates reasoning about threads topology.

What I suggest is to rearrange UI items a bit and explain why:

![](https://i.gyazo.com/20cc97d1f357eccd1150bb52d5927f6c.png)

There can of course be many variants of the solutions to above problems:
- Instead of an arrow, initial message could be differentiated from the new messages by:
  - painting in the yarn color (but that would make the log look like a rainbow - not great)
  - painting it in simply less pronounced color (closer to background - but may become less readable in daylight)
  - changing the background color (like zebra in tables)
  - using smaller font
  - it can itself be shortened
  
- While initial message should be shortened/ellipsized when it's more than one line, new messages may or may not appear in full.

  Personally I would show only the first line of new message: this way I will immediately be able to read answers like "OK" but will have to open the thread to read any in-depth answers, and at the same it would keep the main log not so cluttered by threads.
  
- To show thread starting message or previous message in the thread?

  Starting message will make easier to identify threads IMO, but there's no one size fits all answer.

- Your own ideas :)
  
What do you think @deepbluev7 ? It should be a piece of cake for you to implement, so please consider it :)

**Comment by hiiamboris** [comment]
@deepbluev7 ping

**Comment by deepbluev7** [comment]
Can you create a new issue for it? It is hard to track so far down :)
